### PR TITLE
First prototype of the serialization

### PR DESCRIPTION
**Pull request by jbmouret**

### CODE REVIEW

File: src/limbo/mean/mean.hpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+          \rst
+          Base struct for mean definition.
+          \endrst
+        */
+        template <typename Params>
+        struct BaseMean {
+            BaseMean(size_t dim_out = 1) : _dim_out(dim_out) {}
+
+            template <typename GP>
+            Eigen::VectorXd operator()(const Eigen::VectorXd& v, const GP&) const
```
Comment (jbmouret) [inline src/limbo/mean/mean.hpp]: why keeping it in that case?

File: src/limbo/mean/mean.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                std::cerr << "'BaseMean' should never be called!" << std::endl;
+                assert(false);
+                return Eigen::VectorXd();
+            }
+
+            virtual size_t h_params_size() const { return 0; }
```
Comment (jbmouret) [inline src/limbo/mean/mean.hpp]: Why do we have all these virtuals? there are useless here.

File: src/limbo/mean/mean.hpp (hunk 1 above)
Comment (costashatz) [reply src/limbo/mean/mean.hpp]: > why keeping it in that case?

Not really a reason.. I can remove it..

File: src/limbo/mean/mean.hpp (hunk 2 above)
Comment (costashatz) [reply src/limbo/mean/mean.hpp]: > Why do we have all these virtuals? there are useless here.

True.. I still have OOP microbes in me..!

**Comment by costashatz** [comment]
@jbmouret @Aneoshun I think this PR is ready to be merged.. What do you think?

**Comment by costashatz** [comment]
@jbmouret @Aneoshun I take back my previous comment; we are far from being able to merge!

When the kernel is not `SquaredExpARD` (or when the mean function has parameters; e.g., the `Constant`), the hyperparameters are not saved and they are read from the parameters struct. Thus, if the user does not use the exact same parameters, he/she will not get the same GP as the one he saved. I see the following solution:

We require for all kernel and mean functions to have parameters (even if not optimized). This is generally OK, but then there's no point in having dynamic parameters for these ones (they will only serve as initializations -- this is the case for the `SquaredExpARD` kernel at the moment). Also, that way we will merge the `SquaredExpARD` and the `Exp` kernel finally (which we should have done a long time ago).

What do you think? Do you have any other idea?

**Comment by jbmouret** [comment]
I am not sure to understand what you mean. 

When the parameters are in the struct, they are part of the time, therefore they are not saved (but they are saved in a file during the compilation). We do not need/want to serialize these parameters. We only care about the dynamic part: the hyperparameters & the data.

**Comment by Aneoshun** [comment]
I agree with @jbmouret. The static parameters are saved in the executable. If someone change the executable then it makes sense that previously saved models are not relevant anymore.

**Comment by costashatz** [comment]
After a discussion with JB, we concluded the following:

- We should enable the optimization of all the hyperparameters of the kernel and mean functions that have any. As we saw in our 2 last papers, [1](https://arxiv.org/abs/1709.06919) and [2](https://arxiv.org/abs/1709.06917), being able to optimize the parameters of the mean function (or selecting different mean functions -- mostly equivalent with the previous one) using MLE is a very powerful approach. We can find several examples, where optimizing even the constant mean would be beneficial.
- This has nothing to do with the saving/loading functionality. Nevertheless, since we have this PR open we can include the changes here (since they are changing relevant code).

@Aneoshun What do you think (we can always make a new PR if you don't like having both changes in one PR)?

**Comment by Aneoshun** [comment]
@costashatz I fully agree with your propositions. Does this mean that we have to duplicate every kernel/mean functions to have the ARD variant on?

**Comment by costashatz** [comment]
> @costashatz I fully agree with your propositions. Does this mean that we have to duplicate every kernel/mean functions to have the ARD variant on?

No. I almost have the fixes ready and I will push them in a while.

**Comment by costashatz** [comment]
@Aneoshun @jbmouret Let me know what you think of the changes..

**Comment by Aneoshun** [comment]
I see that the Matern kernels do not have a grad implementation. Shall we do it?

**Comment by costashatz** [comment]
@Aneoshun I added the gradients for the Matern kernels. I also added a test for checking the gradients of all the kernels (via finite differences). Surprisingly, the `SquaredExpARD` with `k>0` fails the check! I have commented out this specific test for now and I will have a look at it tomorrow (if you can look at it a bit also, since you wrote it)..

**Comment by Aneoshun** [comment]
Sure I will

**Comment by costashatz** [comment]
> Surprisingly, the SquaredExpARD with k>0 fails the check

I looked into it and the computations seem correct to me. I do not know why the result is different from the finite differences one.. @Aneoshun any ideas?

**Comment by Aneoshun** [comment]
I spotted an error in the gradient computation. I fixed this, but I cannot commit the fix yet (see my issue in format_code).

The unit-test still fails but significantly less than before and the results look very close. Maybe the 1.e-6 is to severe?

**Comment by costashatz** [comment]
> I fixed this, but I cannot commit the fix yet (see my issue in format_code).

You can remove the [git pre-commit hook](https://github.com/resibots/format_code#pre-commit-hook-in-each-respository) (simply remove the symbolic link; I will handle the formatting later)..

> The unit-test still fails but significantly less than before and the results look very close. Maybe the 1.e-6 is to severe?

Could be. Let me see the results and I can adjust it accordingly..

**Comment by costashatz** [comment]
@Aneoshun I pushed a newer version that it passes the tests. The problem was the discretization used for the finite differences; I was using a too big step for this specific function. Decreasing the step (and thus increasing the accuracy of the finite differences method) solved the issue. Thanks for fixing the error in the gradients.

Now that we are editing the `SquaredExpARD` kernel, I have a question. Why the `A` matrix takes only positive values? Reading the Rassmusen book (pages 106-107), it seems to me that they should lie into [-1,1] rather than being positive (or at least unbounded). What do you think?

**Comment by Aneoshun** [comment]
Excellent question... I think that's a mistake, and we can use any real value (not bunded). In the book, the value 1,-1 is just an example of configuration. 
(we should recompute the gradient in this case...). I can fix this if you want (the math are quite fresh in my mind).

**Comment by costashatz** [comment]
> I can fix this if you want (the math are quite fresh in my mind).

If you can, yes please..

**Comment by Aneoshun** [comment]
Done.

**Comment by costashatz** [comment]
> Done.

Thanks! I fixed a minor Eigen thing and the tests for the kernels!

Once the tests pass, I think we should be ready to merge... Do you see anything missing/that needs a change?

**Comment by jbmouret** [comment]
**OK for merging**

**Pull request merged**